doc/developer/design/20251015_builtin_schema_migration.md

```diff
@@ -0,0 +1,152 @@
+  - If the new schema is backward compatible with the expected schema, it tries to `compare_and_evolve_schema` again.
+  - Otherwise it has been fenced out by a newer version and halts.
+
+During shard replacement, if the `compare_and_append` call on the migration shard fails, the process restarts the shard replacement mechanism, re-reading the migration shard and acting according to its contents.
+Note that here we do nothing to fence out older versions when a new leader has taken over.
+We instead expect either one of the catalog fencing mechanisms or the upgrade orchestration to eventually take care of that.
+
+### Migration Instructions
+
+To avoid data loss and other surprises caused by automatic builtin schema migrations, we introduce the concept of explicit migration instructions.
+A migration instruction instructs the process which builtin collection to migrate at which version, and which mechanism to use.
+
+Migration instructions are kept in a hard-coded list:
+
+```rust
+const MIGRATIONS: &[(Version, SystemObjectDescription, Mechanism)] = &[
+    (
+        V123,
+        SystemObjectDescription {
+            schema_name: "mz_catalog".into(),
+            object_name: "mz_tables".into(),
+            object_type: CatalogItemType::Table,
+        },
+        Mechanism::Evolution,
+    ),
+    ...
+];
+
+enum Mechanism {
+    Evolution,
+    Replacement,
+}
+```
+
+A migration from version X to version Y involves the following steps:
+
+  * Collect from the `MIGRATIONS` list all entries with versions > X and <= Y.
+  * Group the collected entries by the object description.
+  * For each object to migrate, select a mechanism based on the collected `Mechanism` values:
+    * If any value is `Mechanism::Replacement`, select shard replacement.
+    * Otherwise select schema evolution.
+  * For each object to migrate, perform the migration using the selected mechanism.
+
+Note that merging `Evolution` migrations like this is sound because persist requires that each shard schema is backward compatible with any previous schema registered with the shard.
+Which means schema evolution will succeed even if we skip intermediary schemas.
+
+We safeguard against schema changes that don't have an associated `MIGRATIONS` entry by using the existing builtin fingerprint mechanism.
+Each builtin collection has a fingerprint, derived from its schema definition, stored in the catalog.
+By comparing the stored fingerprint with the fingerprint of the object definition in the code, we can detect schema changes that require migration.
+If we detect a schema change but no corresponding `MIGRATIONS` entry, we immediately abort the process.
+
+## Alternatives
+
+As an alternative to schema evolution, we can consider a migration scheme that creates a new shard and copies over all existing data from the old shard, performing the migration in the process.
```

Comment:
Could we also just start hydrating based off the savepoint snapshot of the catalog? Similar to how the upgrade checker checks if the schema is backwards compatible?

An issue here is it wouldn't exactly be zero downtime because we would've missed updates during the upgrade for built in collections. But maybe that's fine if they're just builtin collections?

Reply:
Can you elaborate on what you mean by "start hydrating based off the savepoint snapshot of the catalog"? What persist shards are we hydrating from, the ones that exist in the leader environment or new ones?